[WIP] `async` and `.await` chapter (#3909)
rust-lang/book: src/ch17-00-async-await.md

## Async and Await

The threading-based concurrency model is one of the oldest concurrency models in
computing, and it was present and well-supported in Rust since 1.0. In the past
few decades, though, many programming languages have been experimenting with
other approaches to concurrency, especially asynchronous programming, or
*async*.

It took a few years to work out the right design for async in Rust. After a
bunch of experimentation and design work in the library ecosystem, Rust added
language-level support for async in Rust 1.39, in 2019, and there is a thriving
ecosystem of crates supporting all sorts of interesting capabilities offered by

> those language primitives.

Reviewer: In my opinion, the history is less important to a new Rust programmer than the present. I recommend moving most of this into the end of the chapter or omitting it entirely.

The chapter's introduction should really explain what someone is going to learn and why.

I'm uneasy about entrenching the idea that async Rust is a completely new model. It's based on traits and types, like the rest of the language. The `async` and `.await` keywords don't represent completely new capabilities. They're there to make working with a `Future` easier, in the same way that `?` makes working with `Result` and `Option` easier.

From a grammar point of view, I recommend splitting many of these sentences to shorten them. For example, the sentence "In the past few decades, though, many programming languages have been experimenting with other approaches to concurrency, especially asynchronous programming, or *async*." could be rewritten as "In the past few decades, though, many programming languages have been experimenting with other approaches to concurrency. One that's been especially promising is *asynchronous programming*. Asynchronous programming, or *async* for short, has been part of Rust since 1.39.". 

We have let a phrases slip in here that purport to make a claim, but which are unverifiable, such as "well-supported", "interesting capabilities" and "right design". I'm also wary of jargon in an introductory text book, e.g. "language primitives".

Lastly, I am pretty sure that the facts here are incorrect. Rust's support for threading preceded 1.0 by quite a long way and the original concurrency model was the actor model. The reason why it'll be important to be very accurate with the history is that this book will be cited by many people.

Reviewer 2: File another one under “Chris is the author but not the target audience.” 😂 Concur, and I like @carols10cents’ overall suggestion here.

Reviewer 2: The introduction is probably not all the way there, but I actually just dropped all of the history and rewrote swaths of this introduction entirely this morning (eb41efb3). I think getting rid of the history entirely works well (it’s not actually relevant to readers just wanting to learn this!). I think we probably *do* want a note late in the chapter about how it’s continuing to evolve, but probably in the conclusion, rather than the introduction.

In the rest of this chapter, we will:

* see how to use Rust’s `async` and `.await` syntax
* explore how to use the async model to solve some of the same challenges we
  looked at in Chapter 16
* look at how multithreading and async provide complementary solutions, which
  you can even use together in many cases

First, though, let’s explore what async gives us.

### Why Async?

Many operations we ask the computer to do can take a while to finish. For
example, if you used a video editor to create a video of a family celebration,
exporting it could take anywhere from minutes to hours. Similarly, downloading a
video shared by someone in your family might take a long time. It would be nice
if we could do something else while we are waiting for those long-running
processes to complete.

The video export will use as much CPU and GPU power as it can. If you only had
one CPU core, and your operating system never paused that export until it
completed, you could not do anything else on your computer while it was running.
That would be a pretty frustrating experience, though. Instead, your computer

> can (and does!) invisibly interrupt the export often enough to let you get other

Reviewer: Something to check with the No Starch style guide. The em dash has historically been the "proper" way to indicate a change in the flow.

```suggestion
can—and does!—invisibly interrupt the export often enough to let you get other
```

work done along the way.

> The file download is different. It does not take up very much CPU time. You are
> mostly waiting on data to transfer across the network. You can start reading

Reviewer: ```suggestion
The file download is different. It does not take up very much CPU time. The CPU
needs to wait for data to arrive from the network. While you can start reading
```

Reviewer 2: 👍🏼 Resolving this by way of 10d45b9b (on the #3983 PR and branch)! For the rest of the changes resolved that way, I will just leave the commit hash there for posterity.

from a network socket, but it might take a while for all the data to arrive and

> be fed into the socket by the network controller. Even once the data has all

Reviewer: Few readers will know what a "network controller" is, and I am also wary of "socket".

Reviewer 2: Ah, good call. We actually have mentioned sockets (albeit briefly) already in the book, but more to the point this feels like extraneous detail!

Reviewer 2: 👍🏼 e6692504

arrived, videos can be quite large, so it might take some time to load all the
data from the socket. Maybe it only takes a second or two—but that is a very
long time for a modern processor, which can do billions of operations every
second. It would be nice to be able to put the CPU to use for other work while
waiting for the network call to finish—so, again, your computer will once again
invisibly interrupt your program so other things can happen while the network
operation is still ongoing.

> Note: The video export is the kind of operation which is often described as

> > “CPU-bound”. It is limited by the speed of the computer’s *CPU and GPU*, and

Reviewer: ```suggestion
> “CPU-bound” or “compute-bound”. It is limited by the speed of the computer’s ability to process data within the *CPU* or *GPU*, and
```

Reviewer 2: 👍🏼 e6692504

> how much of that speed it can use. The video download is the kind of operation
> which is often described as “IO-bound,” because it is limited by the speed of
> the computer’s *input and output*. It can only go as fast as the data can be
> sent across the network, which means that it can only go as fast as the data
> can be written to the socket by the network controller.

In both of these examples, the concurrency only happens at the level of a whole
program. The operating system interrupts one program to let other
programs get work done. In many cases, since we understand our programs at a
much more granular level than the operating system does, we can spot lots of
opportunities for concurrency that the operating system cannot see.

For example, if we are building a tool to manage file downloads, we should be
able to write our program in such a way that starting one download does not lock
up the UI, and users should be able to start multiple downloads at the same
time. Many operating system APIs for interacting with network sockets are

> *blocking*, though. That is, the function calls block further progress in the
> program when they are called until they return.

Reviewer: ```suggestion
*blocking*, though. That is, these APIs block the program's progress they the data that 
they are processing is completely ready.
```

I've swapped out "function calls" because you use the phrase "operating system APIs".

Reviewer 2: 👍🏼 e6692504

> Note: This is how *most* function calls work, if you think about it! However,
we normally reserve the term “blocking” for function calls which interact with
files, network sockets, or other resources on the computer, because those are
the places where an individual program would benefit from the operation being
*non*-blocking.

We could use threads to avoid blocking while downloading files, by using a
dedicated thread. But it would be nicer if the call were not blocking in the

> first place.

Reviewer: Sorry to nitpick, but threading doesn't strictly "avoid blocking". But I can't think of an alternative that's as succinct and also technically precise.

Reviewer 3: Maybe something like:

We could avoid blocking our main thread, by spawning a dedicated thread to download each file. But it would be nicer if the call were not blocking in the first place.

Reviewer 2: 👍🏼 e6692504

One way to accomplish that would be to use an API built around callbacks. For
each blocking operation, we could pass in a function to call once the operation
completes:

> ```rust,ignore

Reviewer 4: I don't think it's essential to compare async/await to the other async programming models at the beginning of the chapter. Perhaps you could include it at the end of the chapter?

Reviewer: I'm inclined to agree, but I can definitely see both sides. This section's intended to motivate both async programming and Rust's flavor of async.

Reviewer 2: I ended up just cutting this entirely (eb41efb3) for now—we can always add it back if later reviewers end up confused about “why this approach?”

network_socket.read_non_blocking(|result| {
    // ...
});
```

Or we could register callbacks to run when events happen:

```rust,ignore
network_socket.add_listener(Event::ReadFinished, |event| {
    // ...
});
```

Or we could have our functions return a type with `and_then` method, which in
turn accepts a callback which can do more work of the same sort (Historically,
this was the way that Rust did async):

```rust,ignore
network_socket.read_non_blocking().and_then(|result| {
    /* another non_blocking operation */
});
```

Each of these makes it harder to understand both the control flow and the flow

> of data through the program. You can end up with event handler callbacks

Reviewer: > Each of these makes it harder to understand both the control flow and the flow
of data through the program.

[citation needed]

Generalizations like this will alienate some readers, who very much like the callback model thankyouverymuch.

Reviewer 2: Good call, @timClicks – cut this as part of eb41efb3 as well.

till it finishes unless the whole program stops. <!-- TODO: check whether we
pick back up this thread anywhere! Cancellation is important. -->

What we really want to be able to write is something much more direct, like we
would write in blocking code, but with the benefits of getting the data when it
is available and *not* blocking the rest of the program while waiting for the
data to arrive—something like this:

```rust,ignore,does_not_compile
let data = network_socket.read(&path).await;
println!("{data}");
```

That is exactly what Rust’s async abstraction gives us. It is designed to help
us solve all of these issues. Before we will see how this works in practice,
though, we need to dig a little deeper into the differences between parallelism
and concurrency.

### Parallelism and Concurrency

In the previous chapter we treated parallelism and concurrency as mostly
interchangeable. Now we need to distinguish between them more precisely, because
the differences will show up as we start working:

* *Parallelism* is when operations can happen simultaneously.

* *Concurrency* is when operations can make progress without having to wait for
  all other operations to complete.

> 

Reviewer 4: These definitions seem really similar to me. I always liked the mantra, "parallelism is cooperative and concurrency is competitive". I agree with the example of splitting work between people vs. have one person multi-task, but I don't think this definition really vibes with that example.

Reviewer 2: Heh, the funny thing with that framing is that then you end up having to explain how you can have cooperative multitasking with concurrency. We have ourselves a right and proper terminological mess here! In any case I take you point that these seem quite similar.

That said, [this quite interesting talk](https://www.youtube.com/watch?v=X0GJK3ETrs4) on why Java did user-mode threads instead of async/await) framed it just the same way you did, but with an elaboration that I found helpful, around what the performance measure is: throughput (tasks/time) for concurrency vs. latency for parallelism.

I will chew on this and see if we can find something which is clearer!

Reviewer 2: I agree that we need some diagrams, and I *think* I am starting to develop a decent idea of what they should be like, but getting it right is going to take some work.

I also think that trying to *directly* compare them without getting very jargon-y is hard, so I ended up cutting these and going for *just* the analogies in eb41efb3. I’ll be curious to see how that updated version lands for folks.aI agree that we need some diagrams, and I *think* I am starting to develop a decent idea of what they should be like, but getting it right is going to take some work.

I also think that trying to *directly* compare them without getting very jargon-y is hard, regardless of whether we have the diagrams, so I ended up cutting these and going for *just* the analogies in eb41efb3. I’ll be curious to see how that updated version lands for folks. It may be we need something where we have roughly what those bullet points had as the accompanying text *for the diagrams*.

Reviewer 2: I also added diagrams, with work on them in cfaf187b, 472f152e, and a42f27bd.

## Async and Await

In Chapter 16, we saw one of Rust’s approaches to concurrency: using threads.

> Since Rust 1.39, there has been another option for concurrency: the async-await

Reviewer 2: I think this is 100% a me-ism. 😂 Being consistent with the async book seems good. My inclination is to introduce it something like this:

> Since Rust 1.39, there has been another option for concurrency: asynchronous programming, or async.

Then we can just use “async” going forward after that.

> Since Rust 1.39, there has been another option for concurrency: the async-await

Reviewer 2: Done in cf00f054!

while we are waiting for those long-running processes to complete.

In the previous chapter we treated parallelism and concurrency as
interchangeable. Now we need to distinguish between the two a little more:

* *Parallelism* is when operations can happen simultaneously.

* *Concurrency* is when operations can make progress without having to wait for
  all other operations to complete.

One common analogy for thinking about the difference between concurrency and
parallelism is cooking in a kitchen. Parallelism is like having two cooks: one
working on cooking eggs, and the other working on preparing fruit bowls. Those
can happen at the same time, without either affecting the other. Concurrency is
like having a single cook who can start cooking some eggs, start dicing up some
vegetables to use in the omelette, adding seasoning and whatever vegetables are
ready to the eggs at certain points, and switching back and forth between those
tasks.

(This analogy breaks down if you think about it too hard. The eggs keep cooking
while the cook is chopping up the vegetables, after all. That is parallelism,
not just concurrency! The focus of the analogy is the *cook*, not the food,

> though, and as long as you keep that in mind, it mostly works.)

Reviewer 2: Ooooh, this one is definitely better than the cooks-in-the-kitchen baseline.

Heh: it occurs to me that we could also potentially analogize in very similar ways to *programmers* working on a program! There is definitely a (Distributed) Version Control analogy here. Thought I recognize that might track better to someone “old” like me than it would to a lot of relative newcomers who never lived in a world where you would literally lock a file in the repo and commit changes to it and then unlock it before anyone could operate on it. 😅

I will mull on both.

> runtime we are using, that concurrency may use some degree of parallelism
> under the hood, or it may not. More about async runtimes in a later section!

Consider again the examples of exporting a video file and waiting on the video
file to finish uploading. The video export will use as much CPU and GPU power as
it can. If you only had one CPU core, and your operating system never paused
that export until it completed, you could not do anything else on your computer
while it was running. That would be a pretty frustrating experience, though, so
instead your computer can (and does!) invisibly interrupt the export often
enough to let you get other small amounts of work done along the way.

The file upload is different. It does not take up very much CPU time. Instead,
you are mostly waiting on data to transfer across the network. If you only have
a single CPU core, you might write a bunch of data to a network socket and then
wait for it to finish getting sent by the network controller. You could choose
to wait for all the data to get “flushed” from the socket and actually sent over
the network, but if there is a busy network connection, you might be waiting for

> a while… with your CPU doing not much! Thus, even if you make a blocking call to

Reviewer 2: Ah, good catch. `E_TOO_CLOSE_TO_MATERIAL` I think!

> a while… with your CPU doing not much! Thus, even if you make a blocking call to

Reviewer 2: Fixed this: I now define blocking in the first section, [here](https://github.com/rust-lang/book/pull/3909/files#diff-759f9d1908c9cbc559ae017c2afda23db0cd1e101823202b0749d108f2c292aaR63-R69).

write to a socket, your computer probably does other things while the network
operation is happening.

In both of these cases, it might be useful for *your program* to participate in
the same kind of concurrency the computer is providing for the rest of the
system. One way to do this is the approach we saw last chapter: using threads,
which are provided and managed by the operating system. Another way to get
access to concurrency is using language-specific capabilities—like async-await.

A big difference between the cooking analogy and Rust’s async-await model for
concurrency is that in the cooking example, the cook makes the decision about

> when to switch tasks. In Rust’s async-await model, the tasks are in control of

Reviewer 2: Another reason to hate that analogy! 😂
